### PR TITLE
Skip package-lock.json and yarn.lock in node_modules #431

### LINKED ISSUE #431: Only "top level" lock files should be inspected for NPM packages
Preface: This is on the border of "enhancement" and "bug", depending on how you look at it.

---

Today, when looking for NPM-related lock files (`package-lock.json` and `yarn.lock)`, Syft finds files with these names at any location in the file tree, and then it parses each found file for a list of packages and their versions.

According to the [`package-lock.json` docs](https://docs.npmjs.com/cli/v7/configuring-npm/package-lock-json#package-lockjson-vs-npm-shrinkwrapjson) and the [`yarn.lock` docs](https://classic.yarnpkg.com/en/docs/yarn-lock/#toc-current-package-only), lock files found for a project's dependencies (i.e. not for the project itself) **should be ignored entirely**. In other words, only a project's **_top-level_ lock file** will ever be used by the tooling, and, of particular importance for Syft, during installations and builds for that project.

Today these lock files are used during directory scans and not during image scans. This fact may change in the future, but still, it's worth considering a possible relationship between the implementation of _this issue_ and the implementation of configurable cataloging methods.

Lock files are used for directory scans because the primary use case for directory scans is to scan a software project, typically from the root of a repository.

With this in mind, it could easily be considered **more correct** to surface only the top-level lock files' contents, both for SBOM generation and for downstream vulnerability matching. That is, a project's dependencies — even in the exhaustive sense — **would not** include packages found in lower-level lock files but not found in the top-level lock file. This is important when it comes to creating an SBOM (answering "what software ships with this project?") and identifying vulnerabilities (answering "what vulnerabilities would be present when this code is packaged up and published?").

**IMPORTANT:** While various implementations are possible, implementers should strongly consider the potential for their solution to cause Syft not to find valid packages. The intent of this bug is to reduce false positives, but it would be easily possible to produce a solution that introduces false negatives, which should be avoided if possible.

### PR DESCRIPTION
**Pull request by kzantow**
This fixes #431 and adds integration tests + fixture projects

NOTE: most of these changes are adding a couple "real" javascript projects (I just created them using npm & yarn), I can delete a bunch of the files and still have meaningful tests, if it makes more sense. These are the only files that are non-fixture changes (I wish there was a way to path filter the Github files changed list...):

[syft/pkg/cataloger/javascript/parse_package_lock.go](https://github.com/anchore/syft/pull/485#diff-6d74b1d8c883d795c4330f00b316f33990309c61970bdbe499cf801194da1975)

[syft/pkg/cataloger/javascript/parse_yarn_lock.go](https://github.com/anchore/syft/pull/485/files#diff-29e9c5928f05d10cf6427d487f5eb5d16cb20e75980b291686983ebf74f7d64a)

[test/integration/node_packages_test.go](https://github.com/anchore/syft/pull/485/files#diff-532b65a41aa235c492425b46b2cd9da963b475d673bfd679732f619078d49822)

### CODE REVIEW

File: test/integration/node_packages_test.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+				t.Errorf("found packages from package-lock.json in node_modules: %s", actualLocation)
+			}
+		}
+		foundPackages.Add(actualPkg.Name)
+	}
+
+	// ensure that integration test commonTestCases stay in sync with the available catalogers
+	if len(foundPackages) != 6 {
+		t.Errorf("found the wrong set of npm package-lock.json packages %d != %d", len(foundPackages), 6)
+	}
```
[the remark this answers is not in the document]
Comment (luhring) [reply test/integration/node_packages_test.go:28]: Separate comment: With `%d != %d`: I've found it's more helpful to folks reading the test failure logs to be explicit/obvious about which value is expected vs. actual. For example: `actual: %d, expected: %d`

File: test/integration/node_packages_test.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+				t.Errorf("found packages from yarn.lock in node_modules: %s", actualLocation)
+			}
+		}
+		foundPackages.Add(actualPkg.Name)
+	}
+
+	// ensure that integration test commonTestCases stay in sync with the available catalogers
+	if len(foundPackages) != 5 {
+		t.Errorf("found the wrong set of yarn.lock packages %d != %d", len(foundPackages), 5)
+	}
```
Comment (luhring) [inline test/integration/node_packages_test.go:48]: (same two comments as for analogous section in `TestNpmPackageLockDirectory`)

**Review by luhring — COMMENTED** [review]
Nice fix! 🎉 

I left some thoughts for consideration and discussion...

File: syft/pkg/cataloger/javascript/parse_package_lock.go [hunk 3]
```diff
@@ -27,9 +28,17 @@ type Dependency struct {
...
-func parsePackageLock(_ string, reader io.Reader) ([]pkg.Package, error) {
+// parsePackageLock parses a package-lock.json and returns the discovered JavaScript packages.
+func parsePackageLock(path string, reader io.Reader) ([]pkg.Package, error) {
 	packages := make([]pkg.Package, 0)
+
+	// in the case we find package-lock.json files in the node_modules directories, skip those
+	// as the whole purpose of the lock file is for the specific dependencies of the root project
+	isNodeModulesPath, err := regexp.MatchString("[\\/]node_modules[\\/]", path)
+	if err != nil || isNodeModulesPath {
+		return packages, err
```
[the remark this answers is not in the document]
Comment (kzantow) [reply syft/pkg/cataloger/javascript/parse_package_lock.go:39]: would returning `nil, nil` be a **panic** here, though?: https://github.com/anchore/syft/blob/main/syft/pkg/cataloger/common/generic_cataloger.go#L49-L57

[the remark this answers is not in the document]
Comment (kzantow) [reply syft/pkg/cataloger/javascript/parse_package_lock.go]: Answer: no! :) _**the more you know**_

**Review by luhring — APPROVED** [review]
Beautiful work! :chefs-kiss:

File: syft/pkg/cataloger/javascript/parse_yarn_lock.go [hunk 4]
```diff
@@ -34,9 +34,16 @@ const (
 	noVersion = ""
 )
 
-func parseYarnLock(_ string, reader io.Reader) ([]pkg.Package, error) {
+func parseYarnLock(path string, reader io.Reader) ([]pkg.Package, error) {
 	var packages []pkg.Package
 
+	// in the case we find yarn.lock files in the node_modules directories, skip those
+	// as the whole purpose of the lock file is for the specific dependencies of the project
+	isNodeModulesPath, err := regexp.MatchString("[\\/]node_modules[\\/]", path)
```
[the remark this answers is not in the document]
Comment (kzantow) [reply syft/pkg/cataloger/javascript/parse_yarn_lock.go]: the one note I would have about this is: there _are_ some cases where yarn might do things a bit differently (e.g. [yarn 2 has a new way of managing dependencies called plug-n-play](https://yarnpkg.com/features/pnp)), but this doesn't preclude adding a common method

**Pull request merged**